README.md
[![npm][npm-badge]][npm-badge-url]
[![license][npm-license]][npm-license-url]

[npm-badge]: https://img.shields.io/npm/v/@deckdeckgo/slide-countdown
[npm-badge-url]: https://www.npmjs.com/package/@deckdeckgo/slide-countdown
[npm-license]: https://img.shields.io/npm/l/@deckdeckgo/slide-countdown
[npm-license-url]: https://github.com/deckgo-community/countdown/blob/master/LICENSE

The "Countdown" template displays a countdown until your presentation starts.

It could be handy, for example when you organize a meetup, to display a countdown until the event start.

## Table of contents

- [Installation](#installation)
  - [From a CDN](#from-a-cdn)
  - [From NPM](#from-npm)
  - [Framework integration](#framework-integration)
- [Usage](#usage)
  - [Usage](#usage-1)
  - [Slots](#slots)
- [Attributes](#attributes)
  - [Example without any slots](#example-without-any-slots)
- [Theming](#theming)
- [Development](#development)
- [License](#license)

## Installation

This template could be added to your presentation using the following methods.

### From a CDN

It's recommended to use [unpkg](https://unpkg.com/) if you want to use this template from a CDN. To do so, add the following include script in the main HTML file of your project:

```
<script type="module" src="https://unpkg.com/@deckdeckgo/slide-countdown@latest/dist/deckdeckgo-slide-countdown/deckdeckgo-slide-countdown.esm.js"></script>
```

### From NPM

To install this template in your project from [npm](https://www.npmjs.com/package/@deckdeckgo/slide-countdown) run the following command:

```bash
npm install @deckdeckgo/slide-countdown
```

### Framework integration

The [Stencil documentation](https://stenciljs.com/docs/overview) provide examples of framework integration for [Angular](https://stenciljs.com/docs/angular), [React](https://stenciljs.com/docs/react), [Vue](https://stenciljs.com/docs/vue) and [Ember](https://stenciljs.com/docs/ember).

That being said, commonly, you might either `import` or `load` it:

#### Import

```
import '@deckdeckgo/slide-countdown';
```

#### Loader

```
import { defineCustomElements as deckDeckGoSlideElement } from '@deckdeckgo/slide-countdown/dist/loader';
deckDeckGoSlideElement();
```

## Usage

The "Countdown" slide's Web Component could be integrated using the tag `<deckgo-slide-countdown/>`.

### Usage

```
<deckgo-deck>
    <deckgo-slide-countdown until="2019-08-05T23:25:59.000+02:00">
        <h1 slot="title">My presentation starts at</h1>
        <p slot="days">Days</p>
        <p slot="hours">Hours</p>
        <p slot="minutes">Minutes</p>
        <p slot="seconds">Seconds</p>
    </deckgo-slide-countdown>
</deckgo-deck>
```

### Slots

The slots `title` as well as `days`, `hours`, `minutes` and `seconds` are optional.

## Attributes

The time until your presentation should be provided to render the countdown. This value could be either passed through attributes `days`, `hours`, `minutes` and `seconds` or as a particular date using `until`.

This component exposes the following attributes:

| Attribute         | Type    | Default | Description                                                                                                     |
| ----------------- | ------- | ------- | --------------------------------------------------------------------------------------------------------------- |
| days              | number  | 0       | The amount of days before your presentations (max. 99 will be displayed)                                        |
| hours             | number  | 0       | The amount of hours before your presentations (max. 23)                                                         |
| minutes           | number  | 0       | The amount of minutes before your presentations (max. 59)                                                       |
| seconds           | number  | 0       | The amount of seconds before your presentations (max. 59)                                                       |
| until             | string  |         | A specific date and time until when your presentation will start                                                |
| custom-background | boolean | false   | If you would provide a background for the all deck and a specific one for this slide, set this option to `true` |
| custom-actions    | boolean | false   | If you would provide actions for the all deck and a specific one for this slide, set this option to `true`      |

If you provide a date using `until`, note that the format should be provided as a valid and parsable date. See [Date.parse()](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Date/parse) for more information about the format.

### Example without any slots

```
<deckgo-deck>
  <deckgo-slide-countdown hours="0" minutes="10" seconds="45">
  </deckgo-slide-countdown>
</deckgo-deck>
```

## Theming

The following theming options will affect this component if set on its host or parent.

| CSS4 variable                              | Default                                                                   | Note                                                        |
| ------------------------------------------ | ------------------------------------------------------------------------- | ----------------------------------------------------------- |
| --background                               |                                                                           |                                                             |
| --color                                    |                                                                           |                                                             |
| --slide-padding-top                        | 16px                                                                      | Padding top of the all slide                                |
| --slide-padding-end                        | 32px                                                                      | Padding right of the all slide                              |
| --slide-padding-bottom                     | 16px                                                                      | Padding bottom of the all slide                             |
| --slide-padding-start                      | 32px                                                                      | Padding left of the all slide                               |
| --slide-countdown-container-padding-bottom | 8%                                                                        | The bottom padding of the displayed time container          |
| --slide-countdown-digits-max-width         | 36em                                                                      | The max width of the container containing all digits        |
| --slide-countdown-digits-minmax-width      | 12em                                                                      | The grid minmax value of one of the three columns of digits |
| --slide-countdown-digits-width             | 4em                                                                       | The width of one of the (two) digits                        |
| --slide-countdown-digits-height            | 4m                                                                        | The height of the (two) digits                              |
| --slide-countdown-digits-background        |                                                                           | The background color of the (two) digits                    |
| --slide-countdown-digits-border-radius     |                                                                           | The border-radius of the (two) digits                       |
| --slide-countdown-digits-box-shadow        | 0 3px 4px 0 rgba(0, 0, 0, .2), inset 2px 4px 0 0 rgba(255, 255, 255, .08) | The box-shadow of the (two) digits                          |
| --slide-countdown-digit-margin-right       | 0.625em                                                                   | The space between two digits                                |
| --slide-countdown-digits-font-size         | 3em                                                                       | The border-radius of the (two) digits                       |
| --slide-countdown-digits-font-weight       |                                                                           | The font-weight of the (two) digits                         |
| --slide-countdown-digits-color             |                                                                           | The color of the (two) digits                               |

## Development

To develop and run this Web Component locally, proceed as following:

```
git clone https://github.com/deckgo-community/countdown
cd countdown
npm install
npm run start
```

## License

MIT © [David Dal Busco](mailto:david.dalbusco@outlook.com), [Nicolas Mattia](mailto:nicolas@nmattia.com) & [Gaurav Mahto](https://github.com/gauravmahto)

[deckdeckgo]: https://deckdeckgo.com


CHANGELOG.md
# 2.4.0 (2021-05-26)

### Build

- update dependencies including `slide-utils`

# 2.3.0 (2020-01-22)

### Features

- repo moved
- properties reflected as attributes
- reset timer on properties changes
- default padding-bottom modified to `8%`

# 2.2.2 (2020-11-14)

### Fix

- cross-browser user selection

# 2.2.1 (2020-11-14)

### Fix

- publish to npm

# 2.2.0 (2020-11-14)

### Style

- default user select set to `text` instead of `none`

# 2.1.3 (2020-11-08)

### Features

- update slide-utils for lazy loading images

# 2.1.2 (2020-10-29)

### Fix

- update slide-utils for word-cloud

# 2.1.1 (2020-10-23)

### Fix

- es5 has been dropped with v2, update `unpkg` reference

# 2.1.0 (2020-09-24)

### Features

- support for `vertical` and `papyrus`
- update dev dependencies

# 2.0.0 (2020-09-03)

### Breaking Changes

- IE11, Edge 16-18 and Safari 10 not supported

# 1.1.2 (2020-07-12)

### Features

- pointer-events supported for `header` and `footer`

# 1.1.1 (2020-07-11)

### Style

- adjusted `header` and `footer` positions

# 1.1.0 (2020-07-10)

### Features

- add `header` and `footer`

### Refactoring

- remove `custom-actions` and `custom-background` properties because these are only use as attributes

# 1.0.1 (2020-05-11)

### Features

- update Stencil for Gatsby build

# 1.0.0 (2020-03-19)

To infinity and beyond 🚀

### Features

- update dependencies


deckdeckgo.desc.ts
import {JsonDocs} from '@stencil/core/internal';
import {JsonDocsComponent, JsonDocsProp, JsonDocsSlot} from '@stencil/core/internal/stencil-public-docs';

import * as fs from 'fs';

interface DeckDeckGoAuthor {
  name?: string;
  url?: string;
}

interface DeckDeckGoSlot {
  name: string;
  placeholder?: string;
  types?: string[];
  author?: DeckDeckGoAuthor;
}

interface DeckDeckGoProp {
  name: string;
  type: string;
  placeholder?: string;
}

const parseSlots = (slots: JsonDocsSlot[] | undefined): DeckDeckGoSlot[] | undefined => {
  if (!slots || slots.length <= 0) {
    return undefined;
  }

  return slots.map((slot: JsonDocsSlot) => {
    const docs: string[] | undefined = slot.docs?.split('-');
    const types: string[] | undefined = docs && docs.length >= 2 ? docs[1]?.trim().split(',') : undefined;

    return {
      name: slot.name,
      ...(docs && docs.length >= 1 && docs[0] !== '' && {placeholder: docs[0].trim()}),
      ...(types && types.length > 0 && {types})
    };
  });
};

const parseProps = (props: JsonDocsProp[] | undefined): DeckDeckGoProp[] | undefined => {
  if (!props || props.length <= 0) {
    return undefined;
  }

  return props
    .filter((prop: JsonDocsProp) => prop.reflectToAttr && ['string', 'number', 'boolean'].includes(prop.type))
    .map((prop: JsonDocsProp) => {
      return {
        name: prop.attr,
        type: prop.type,
        ...(prop.docs && {placeholder: prop.docs})
      };
    });
};

const parseAuthor = (): DeckDeckGoAuthor | undefined => {
  const packageJson = require('./package.json');

  if (!packageJson?.author?.name) {
    return undefined;
  }

  return {
    name: packageJson.author.name,
    ...(packageJson.author.url && {url: packageJson.author.url})
  };
};

export const generateDesc = (docs: JsonDocs) => {
  if (!docs || !docs.components) {
    console.warn('No docs or components provided.');
  }

  const author: DeckDeckGoAuthor | undefined = parseAuthor();

  const components = docs.components.map((cmp: JsonDocsComponent) => {
    const slots: DeckDeckGoSlot[] | undefined = parseSlots(cmp.slots);
    const props: DeckDeckGoProp[] | undefined = parseProps(cmp.props);

    const propType: DeckDeckGoProp | undefined = props?.find((prop: DeckDeckGoProp) => prop.name === 'type');
    if (propType) {
      throw new Error('The property "type" is a reserved keyword. Rename it to resolve the issue.');
    }

    return {
      tag: cmp.tag,
      ...(author && {author}),
      ...(props && {props}),
      ...(slots && {slots})
    };
  });

  fs.writeFileSync('./src/components.desc.json', JSON.stringify(components, null, 2));
};


src/components.d.ts
/* eslint-disable */
/* tslint:disable */
/**
 * This is an autogenerated file created by the Stencil compiler.
 * It contains typing information for all components that exist in this project.
 */
import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";
export namespace Components {
    interface DeckgoSlideCountdown {
        "afterSwipe": () => Promise<void>;
        "beforeSwipe": (_enter: boolean, _reveal: boolean) => Promise<boolean>;
        /**
          * The amount of days before your presentations
         */
        "days": number;
        "hideContent": () => Promise<void>;
        /**
          * The amount of hours before your presentations (max. 23)
         */
        "hours": number;
        "lazyLoadContent": () => Promise<void>;
        /**
          * The amount of minutes before your presentations (max. 59)
         */
        "minutes": number;
        "revealContent": () => Promise<void>;
        "seconds": number;
        "start": () => Promise<void>;
        "stop": () => Promise<void>;
        /**
          * A specific date and time until when your presentation will start (format: 2021-08-27T23:25:59.000+02:00)
         */
        "until": string;
    }
}
declare global {
    interface HTMLDeckgoSlideCountdownElement extends Components.DeckgoSlideCountdown, HTMLStencilElement {
    }
    var HTMLDeckgoSlideCountdownElement: {
        prototype: HTMLDeckgoSlideCountdownElement;
        new (): HTMLDeckgoSlideCountdownElement;
    };
    interface HTMLElementTagNameMap {
        "deckgo-slide-countdown": HTMLDeckgoSlideCountdownElement;
    }
}
declare namespace LocalJSX {
    interface DeckgoSlideCountdown {
        /**
          * The amount of days before your presentations
         */
        "days"?: number;
        /**
          * The amount of hours before your presentations (max. 23)
         */
        "hours"?: number;
        /**
          * The amount of minutes before your presentations (max. 59)
         */
        "minutes"?: number;
        "onSlideDidLoad"?: (event: CustomEvent<void>) => void;
        "seconds"?: number;
        /**
          * A specific date and time until when your presentation will start (format: 2021-08-27T23:25:59.000+02:00)
         */
        "until"?: string;
    }
    interface IntrinsicElements {
        "deckgo-slide-countdown": DeckgoSlideCountdown;
    }
}
export { LocalJSX as JSX };
declare module "@stencil/core" {
    export namespace JSX {
        interface IntrinsicElements {
            "deckgo-slide-countdown": LocalJSX.DeckgoSlideCountdown & JSXBase.HTMLAttributes<HTMLDeckgoSlideCountdownElement>;
        }
    }
}
